service/history/historyEngine.go

```diff
@@ -1107,6 +1111,20 @@ func (c *decisionBlobSizeChecker) failWorkflowIfBlobSizeExceedsLimit(blob []byte
 	return true, nil
 }
 
+// Before closing workflow, if there is a in-flight decision, fail the decision first. So that we don't close the workflow with buffered events
+func failInFlightDecisionToClearBufferedEvents(msBuilder mutableState) error {
+	if msBuilder.HasInFlightDecisionTask() {
+		di, _ := msBuilder.GetInFlightDecisionTask()
+
+		event := msBuilder.AddDecisionTaskFailedEvent(di.ScheduleID, di.StartedID, workflow.DecisionTaskFailedCauseResetWorkflow, nil,
+			identityHistoryService, decisionFailureForBuffered, "", "", 0)
+		if event == nil {
+			return &workflow.InternalServiceError{Message: "Failed to add decision failed event."}
+		}
+	}
```

Comment:
should also flush the buffer here, or otherwise, buffered events are still in the buffer